File: s3/s3.go
package s3

import (
	"context"
	"log"
	"os"

	"github.com/minio/minio-go/v7"
	"github.com/minio/minio-go/v7/pkg/credentials"
)

const S3EndpointEnv = "S3_ENDPOINT"
const S3AccessKeyIdEnv = "S3_ACCESS_KEY_ID"
const S3SecretAccessKeyEnv = "S3_SECRET_ACCESS_KEY"

const BucketName = "artifacts"
const DefaultRegion = "us-east-1"

const ObjectNotExistErrorCode = "NoSuchKey"

var client *minio.Client
var ctx context.Context

func Setup() {
	c, err := createClient()
	if err != nil {
		panic("")
	}
	client = c

	ctx = context.Background()
}

func UploadFile(filePath string, objectName string) error {
	contentType := "application/octet-stream"
	ctx := context.Background()
	_, err := client.FPutObject(ctx, BucketName, objectName, filePath, minio.PutObjectOptions{ContentType: contentType})
	if err != nil {
		log.Fatalln(err)
	}
	return nil
}

func IsObjectExists(objName string, checksum string) (bool, error) {
	o := minio.ObjectAttributesOptions{}
	attr, err := client.GetObjectAttributes(ctx, BucketName, objName, o)
	if err != nil {
		if minio.ToErrorResponse(err).Code == ObjectNotExistErrorCode {
			return false, nil
		}
		return false, err
	}
	match := attr.Checksum.ChecksumSHA1 == checksum
	return match, nil
}

func CreateBucketIfNotExists() error {
	exists, err := client.BucketExists(ctx, BucketName)
	if err != nil {
		return err
	}
	if exists {
		log.Printf("Bucket '%s' is alredy exists", BucketName)
		return nil
	}
	err = client.MakeBucket(ctx, BucketName, minio.MakeBucketOptions{Region: DefaultRegion})
	if err == nil {
		log.Printf("Bucket '%s' is created\n", BucketName)
	}
	return err
}

func createClient() (*minio.Client, error) {
	endpoint := os.Getenv(S3EndpointEnv)
	accessKeyID := os.Getenv(S3AccessKeyIdEnv)
	secretAccessKey := os.Getenv(S3SecretAccessKeyEnv)
	useSSL := true

	client, err := minio.New(endpoint, &minio.Options{
		Creds:  credentials.NewStaticV4(accessKeyID, secretAccessKey, ""),
		Secure: useSSL,
	})
	return client, err
}

File: art_keeper/s3.py
from minio import Minio
from minio.error import S3Error


class S3Client:
    def __init__(
        self,
        endpoint: str,
        access_key: str,
        secret_key: str,
        region: str = "us-east-1",
    ):
        self._client = Minio(
            endpoint=endpoint,
            access_key=access_key,
            secret_key=secret_key,
            region=region,
        )

    def upload_file(self, bucket_name: str, object_name: str, file_path: str):
        self._client.fput_object(bucket_name, object_name, file_path)

    def is_object_exists(
        self, bucket_name: str, object_name: str, checksum: str
    ) -> bool:
        try:
            stats = self._client.stat_object(bucket_name, object_name)
            return stats.etag == checksum
        except S3Error:
            return False

    def create_bucket_if_needed(self, name: str):
        found = self._client.bucket_exists(bucket_name=name)
        if not found:
            self._client.make_bucket(bucket_name=name)
            print(f"Created bucket {name}")
        else:
            print(f"Bucket {name} already exists")
